net/libfab/libfab.c

```diff
@@ -361,261 +351,52 @@ M0_INTERNAL void m0_net_libfab_fini(void)
-			return M0_ERR(-EPROTO);
-		for (i = 0; i < ARRAY_SIZE(fab_autotm); ++i) {
-			if (fab_autotm[i] == 0) {
-				tmid = i;
-				break;
-			}
-		}
-		if (i == ARRAY_SIZE(fab_autotm))
-			return M0_ERR(-EADDRNOTAVAIL);
-	}
-	
-	if (pid != 12345)
-		return M0_ERR(-EPROTO);
-	/*
-	 * Deterministically combine portal and tmid into a unique 16-bit port
-	 * number (greater than 1024). Tricky.
-	 *
-	 * Port number is, in binary: tttttttttt1ppppp, that is, 10 bits of tmid
-	 * (which must be less than 1024), followed by a set bit (guaranteeing
-	 * that the port is not reserved), followed by 5 bits of (portal - 30),
-	 * so that portal must be in the range 30..61.
-	 *
-	 * if (tmid >= 1024 || (portal - 30) >= 32)
-	 * 	return M0_ERR_INFO(-EPROTO,
-	 * 		"portal: %u, tmid: %u", portal, tmid);
-	 */
-
-	if (portal < 30)
-		portal = 30 + portal;
-
-	portnum  = tmid | (1 << 10) | ((portal - 30) << 11);
-	M0_ASSERT(portnum < 65536);
-	sprintf(port, "%d", (int)portnum);
-	fab_autotm[tmid] = 1;
-	return M0_RC(0);
-}
-
-/**
- * This function decodes the socket format address and extracts the ip address
- * and port number from it. The socket address format is of the type
- *    family:type:ipaddr[@port]
- *    for example: "inet:stream:lanl.gov@23",
- *                 "inet6:dgram:FE80::0202:B3FF:FE1E:8329@6663" or
- *                 "unix:dgram:/tmp/socket".
- */
-static int libfab_ep_addr_decode_native(const char *ep_name, char *node,
-				        size_t nodeSize, char *port,
-				        size_t portSize, uint8_t *addr_frmt)
-{
-	int   shift = 0;
-	int   f;
-	int   s;
-	int   rc;
-	char *at;
-	char  ip[LIBFAB_ADDR_LEN_MAX] = {};
-
-	for (f = 0; f < ARRAY_SIZE(protf); ++f) {
-		if (protf[f] != NULL) {
-			shift = strlen(protf[f]);
-			if (strncmp(ep_name, protf[f], shift) == 0)
-				break;
-		}
-	}
-
-	if (ep_name[shift] != ':')
-		return M0_ERR(-EINVAL);
-
-	ep_name += shift + 1;
-	for (s = 0; s < ARRAY_SIZE(socktype); ++s) {
-		if (socktype[s] != NULL) {
-			shift = strlen(socktype[s]);
-			if (strncmp(ep_name, socktype[s], shift) == 0)
-				break;
-		}
-	}
-
-	if (ep_name[shift] != ':')
-		return M0_ERR(-EINVAL);
-	ep_name += shift + 1;
-	at = strchr(ep_name, '@');
-	if (at == NULL)
-		return M0_ERR(-EINVAL);
-
-	if (++at == NULL)
-		return M0_ERR(-EINVAL);
-	M0_ASSERT(portSize >= strlen(at) + 1);
-	memcpy(port, at, strlen(at) + 1);
-
-	rc = libfab_hostname_to_ip((char *)ep_name, ip, addr_frmt);
-	if (rc == 0)
-		strcpy(node, ip);
-
-	/* Ignore the error due to gethostbyname() as it will be retried. */
-	return rc >= 0 ? M0_RC(0) : M0_ERR(rc);
+	if (addr->nip_format == M0_NET_IP_LNET_FORMAT)
+		inet_ntop(AF_INET, &addr->nip_ip_n.sn[0], ip,
+			  LIBFAB_ADDR_LEN_MAX);
+	else if (addr->nip_fmt_pvt.ia.nia_family == M0_NET_IP_AF_INET)
+		inet_ntop(AF_INET, &addr->nip_ip_n.sn[0], ip,
+			  LIBFAB_ADDR_LEN_MAX);
+	else if (addr->nip_fmt_pvt.ia.nia_family == M0_NET_IP_AF_INET6)
+		inet_ntop(AF_INET6, &addr->nip_ip_n.ln[0], ip,
+			  LIBFAB_ADDR_LEN_MAX);
```

Comment:
To give priority to Libfabric IPv4 related we can do this for optimization 

if( (addr->nip_fmt_pvt.ia.nia_family == M0_NET_IP_AF_INET) ||  
    (addr->nip_format == M0_NET_IP_LNET_FORMAT) )
     inet_ntop(AF_INET, &addr->nip_ip_n.sn[0], ip,
	 	  LIBFAB_ADDR_LEN_MAX);      
else if (addr->nip_fmt_pvt.ia.nia_family == M0_NET_IP_AF_INET6)
		inet_ntop(AF_INET6, &addr->nip_ip_n.ln[0], ip,
			  LIBFAB_ADDR_LEN_MAX);
		inet_ntop(AF_INET6, &addr->nip_ip_n.ln[0], ip,
			  LIBFAB_ADDR_LEN_MAX);		
else
		M0_LOG(M0_ERROR, "Family is not supported.");	

?

Reply:
Done.